internal/zinc-compile-core/src/main/scala/sbt/internal/inc/HashUtil.scala

```diff
@@ -13,13 +13,24 @@ package sbt
 package internal
 package inc
 
-import java.nio.file.{ Files, Path }
 import net.openhft.hashing.LongHashFunction
 
+import java.nio.file.{Files, Path}
+
 object HashUtil {
   def farmHash(bytes: Array[Byte]): Long =
     LongHashFunction.farmNa().hashBytes(bytes)
 
-  def farmHash(path: Path): Long =
-    farmHash(Files.readAllBytes(path))
+  def farmHash(path: Path): Long = {
+    import sbt.io.Hash
+
+    //allocating many byte arrays for large files may lead to OOME
+    //but it is more efficient for small files
+    val largeFileLimit = 10 * 1024 * 1024
+
+    if (Files.size(path) < largeFileLimit)
+      farmHash(Files.readAllBytes(path))
+    else
+      farmHash(Hash(path.toFile))
```

Comment:
Would this basically stop checking for hash? I don't know if 10MB is good enough threshold for this if for example JAR could change with the same file name (SNAPSHOT etc). Maybe include file size since you grabbed it?

Reply:
Also I am not sure if grabbing file name in general is a good idea since the point of using content hash (as opposed to using last modified) was to make this machine independent and repeatable, and the absolute path could change from machine to machine.

Reply:
`sbt.io.Hash.apply` computes SHA of a file content using DigestInputStream. Maybe it's worth to copy implementation here to be more explicit.

Reply:
My mistake. So the difference basically would be that this would use SHA instead of Farmhash.